### PR TITLE
ffmpeg_4: almost drop

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

Bump almost all of the remaining dependencies on FFmpeg 4, and add a comment about its legacy status so that we can discourage further proliferation and hopefully remove it at some point.

This PR is stacked on top of the following others; please ensure any remaining PRs are reviewed and merged before landing this one:

* https://github.com/NixOS/nixpkgs/pull/327642
* https://github.com/NixOS/nixpkgs/pull/330648
* https://github.com/NixOS/nixpkgs/pull/330657
* https://github.com/NixOS/nixpkgs/pull/331514
* https://github.com/NixOS/nixpkgs/pull/331522
* https://github.com/NixOS/nixpkgs/pull/331558
* https://github.com/NixOS/nixpkgs/pull/331568
* https://github.com/NixOS/nixpkgs/pull/331615
* https://github.com/NixOS/nixpkgs/pull/331623
* https://github.com/NixOS/nixpkgs/pull/331628
* https://github.com/NixOS/nixpkgs/pull/331634
* https://github.com/NixOS/nixpkgs/pull/331636
* https://github.com/NixOS/nixpkgs/pull/331864
* https://github.com/NixOS/nixpkgs/pull/331872
* https://github.com/NixOS/nixpkgs/pull/332035
* https://github.com/NixOS/nixpkgs/pull/332051
* https://github.com/NixOS/nixpkgs/pull/332061
* https://github.com/NixOS/nixpkgs/pull/332111
* https://github.com/NixOS/nixpkgs/pull/336395
* https://github.com/NixOS/nixpkgs/pull/336077
* https://github.com/NixOS/nixpkgs/pull/336110
* https://github.com/NixOS/nixpkgs/pull/336141
* https://github.com/NixOS/nixpkgs/pull/336157
* https://github.com/NixOS/nixpkgs/pull/336164
* https://github.com/NixOS/nixpkgs/pull/336169
* https://github.com/NixOS/nixpkgs/pull/336173
* https://github.com/NixOS/nixpkgs/pull/336177
* https://github.com/NixOS/nixpkgs/pull/336179
* https://github.com/NixOS/nixpkgs/pull/336185
* https://github.com/NixOS/nixpkgs/pull/336187
* https://github.com/NixOS/nixpkgs/pull/336193
* https://github.com/NixOS/nixpkgs/pull/336201
* https://github.com/NixOS/nixpkgs/pull/336410 (by @pabloaul)

All of the package changes that are in any way tricky or potentially controversial were split out into the above PRs. This home straight consists entirely of trivial changes of FFmpeg pins, bumps between upstream versions of the same stability level, and clean backports of simple patches that have already been committed upstream. Everything here has been confirmed to build.

After this PR and the linked ones, the remaining packages depending on `ffmpeg_4` will be:

* `reaper` (binary package; dependent on upstream’s whims)

* `spotify` (binary package; dependent on upstream’s whims)

* `rocmPackages_{5,6}.mivisionx` (both broken for a long time; maybe they’ll update before they become unbroken!)

This turned out to be a bigger project than I thought it would be when I started out, but 34 PRs later, I’m glad to finally see it (almost) completed.

Once all of these PRs are merged, I am going to bump `ffmpeg` to 7 on `staging`, and the real fun can begin.

Result of `nixpkgs-review` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>qstopmotion</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>alephone-apotheosis-x</li>
  </ul>
</details>
<details>
  <summary>27 packages built:</summary>
  <ul>
    <li>alephone</li>
    <li>alephone-durandal</li>
    <li>alephone-eternal</li>
    <li>alephone-evil</li>
    <li>alephone-infinity</li>
    <li>alephone-marathon</li>
    <li>alephone-pathways-into-darkness</li>
    <li>alephone-pheonix</li>
    <li>alephone-red</li>
    <li>alephone-rubicon-x</li>
    <li>alephone-yuge</li>
    <li>alephone.icons</li>
    <li>ardour</li>
    <li>ardour_7</li>
    <li>electricsheep</li>
    <li>guvcview</li>
    <li>libsForQt5.ffmpegthumbs</li>
    <li>octavePackages.video</li>
    <li>olive-editor</li>
    <li>pangolin</li>
    <li>rustplayer</li>
    <li>sumo</li>
    <li>ultrastardx</li>
    <li>untrunc-anthwlock</li>
    <li>vulkan-cts</li>
    <li>webcamoid</li>
    <li>xjadeo</li>
  </ul>
</details>

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>alephone-apotheosis-x</li>
  </ul>
</details>
<details>
  <summary>30 packages built:</summary>
  <ul>
    <li>alephone</li>
    <li>alephone-durandal</li>
    <li>alephone-eternal</li>
    <li>alephone-evil</li>
    <li>alephone-infinity</li>
    <li>alephone-marathon</li>
    <li>alephone-pathways-into-darkness</li>
    <li>alephone-pheonix</li>
    <li>alephone-red</li>
    <li>alephone-rubicon-x</li>
    <li>alephone-yuge</li>
    <li>alephone.icons</li>
    <li>ardour</li>
    <li>ardour_7</li>
    <li>electricsheep</li>
    <li>guacamole-server</li>
    <li>guvcview</li>
    <li>libsForQt5.ffmpegthumbs</li>
    <li>megasync</li>
    <li>octavePackages.video</li>
    <li>olive-editor</li>
    <li>pangolin</li>
    <li>qstopmotion</li>
    <li>rustplayer</li>
    <li>sumo</li>
    <li>ultrastardx</li>
    <li>untrunc-anthwlock</li>
    <li>vulkan-cts</li>
    <li>webcamoid</li>
    <li>xjadeo</li>
  </ul>
</details>

Result of `nixpkgs-review` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>sumo</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>octavePackages.video</li>
    <li>pangolin</li>
    <li>untrunc-anthwlock</li>
  </ul>
</details>

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/kde/ffmpegthumbs.nix
Comment (K900) [inline pkgs/applications/kde/ffmpegthumbs.nix:1]: nit: this is libsForQt5.ffmpegthumbs, not kdePackages.ffmpegthumbs.

Comment (emilazy) [reply pkgs/applications/kde/ffmpegthumbs.nix:1]: Ah, hence the file path…

FIxed, thank you.

**Comment by emilazy** [comment]
This is now just pending on the `libresample` upstream cutting a release with my merged Meson build system patch and hopefully the other PR to fix the tests, and figuring out what to do about `tvheadend`.

If I don’t hear back from the `libresample` upstream within the next couple of days, I’ll rebase my PR on top of the Git HEAD and keep the patch to fix the tests, and I can come back and bump it whenever there’s a proper tag. Similarly, if there’s no response to https://github.com/NixOS/nixpkgs/pull/332259#issuecomment-2305196447 I’ll merge https://github.com/NixOS/nixpkgs/pull/336395. Then this will be ready and I can make the FFmpeg 7 switch happen.

**Review by SigmaSquadron — COMMENTED** [review]
Should the [24.11 release notes](https://github.com/NixOS/nixpkgs/blob/cbdb6ac0e6389f89acd02e3dd0cd3090c67c1a0e/nixos/doc/manual/release-notes/rl-2411.section.md?plain=1#L325) be edited to no longer recommend pinning `ffmpeg_4`?

**Comment by emilazy** [comment]
I’m planning to fold the FFmpeg 5 removal note in to the one about FFmpeg 7 becoming the default when I do that PR, since otherwise I’d be writing something about FFmpeg versions just to rewrite it again a few days from now. (And, anyway, we’ll still have a couple binary packages pinning `ffmpeg_4`, and it’s not the end of the world if out‐of‐tree users do it, so long as they know it’s on borrowed time, so I think “if necessary” is about strong enough as it stands.)

**Comment by emilazy** [comment]
With all the other PRs dealt with, this is now ready for review and merge 🎉

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by Atemu** [comment]
@SuperSandro2000 could you not unassign maintainers of the package the PR is about and merge without notice? Thanks.

**Comment by emilazy** [comment]
@SuperSandro2000 I am also one of the FFmpeg maintainers, precisely as a result of this months‐long project to get rid of old versions. You think I accidentally pinged my co‐maintainers to the latest PR in my long‐running FFmpeg project? I’m sorry, but assuming I made some beginner’s mistake in asking for review here with the amount of care and effort I’ve put into this project is just incredibly condescending. (Well, okay, yes: it was accidental that I added them as assignees and not reviewers, but what does that matter? If that’s a big deal, you could remove them as assignees and add them back as reviewers instead.)

It’s very much about FFmpeg – look at the title! – and contains [an FFmpeg policy decision](https://github.com/NixOS/nixpkgs/pull/336401/commits/778806c4a726c27d8cb631a4df1c74769a4c1a6a) that I specifically wanted their sign‐off on, not to mention my talk of potentially *dropping one of the FFmpeg packages* in the comment immediately before your merge without comment. Did you even read the PR? I’ve been coordinating this whole project with them. Can you just admit you screwed up here rather than deflecting and making excuses…?

@Atemu @jopejoe1 I am sorry about this, and of course still open to any feedback here.

**Comment by SuperSandro2000** [comment]
Oh well derp sorry. 
You seem undecided whether to drop off ffmpeg4-full already in this PR and since you wrote before that this is ready to be merged, I just did that.

**Comment by Atemu** [comment]
Sandro, all of us have been Nixpkgs maintainers for many years. We know how things work around here. It's true that we don't use assignment often (we really should) but if we do, the meaning should be very obvious: The people assigned are responsible for whatever they're assigned to and thereby "own" it. In this case, Emily didn't appear to technically mean to do that but something slightly different but the intention would have been the same.  
If someone else is assigned to some task at your $work, you don't go around and take over their task without asking either, do you?

I think that we should make these rules more explicit for clarity but don't you think it's at least worth *asking* when a situation is, in your opinion, unclear?  

As a maintainer, I feel disregarded here and I regret showing support of restoring your merge privilege a little.

For the record, I generally agree that we should depend on ffmpeg_4 as little as possible but I wouldn't go as far as banning its use in source packages. I'd prefer if people attempted to patch their packages for compat but it shouldn't block the package's inclusion (SHOULD) as I think that, while it's reasonable to *ask* people to patch their package in such a way, you cannot *expect* them to do so. Emily went above and beyond to patch these packages in a way we cannot expect any maintainer to do so IMHO.

**Comment by emilazy** [comment]
I can agree that it’s more of a SHOULD than a MUST. That said, in my experience the packages that actually needed patching (rather than just updating or undoing a pin) were almost universally ones that were pretty crusty and hadn’t seen any upstream activity in many years. (The exceptions were mostly packages that seemed to be moving away from using FFmpeg at all, or stuff like Qt WebEngine – though even there it was a symptom of WebEngine being based on an old Chromium and needing the FFmpeg 7 patches from Chromium upstream backporting.)

My assessment was that it would be questionable to add those packages that needed non‐trivial work in 2024. We couldn’t add a package that depends on versions of FFmpeg *older* than 4 without patching, but even with patching I think it would give us pause – do we really want to add something to the tree that it seems clear off the bat is not going to get any upstream maintenance to deal with any future issues? As a result it seemed unlikely to me that there’d be a compelling new source package that can’t support FFmpeg 6+, and that we shouldn’t add more baggage that will make it harder to drop the version when the lagging binary packages finally catch up.

I’m happy to reword the message if you think I could capture those nuances better, or discuss it further if there’s still disagreement.

**Comment by emilazy** [comment]
Another bit of colour I should add is that many of the packages could easily have been bumped from FFmpeg 4 to FFmpeg 6 with just an unpinning and perhaps an update – part of why this was so much effort is that I tried to get things on the newer FFmpeg 7 if possible to smooth over the upcoming default version migration rather than turning 6 into the new 4. So since depending on FFmpeg 6 for a new package is reasonable (if perhaps suboptimal), even a large portion of the packages that required some active work on my part would have been able to be added without any issues or patching.

**Comment by Letgamer** [comment]
I just ran `strace` on the latest Spotify Version:
```bash
strace -f -e trace=openat,openat2 /nix/store/f73parrf71pqhnxmwhf73jjhg776dlr3-spotify-1.2.92.147.g5b8f9367/share/spotify/spotify 2> /tmp/spotify-strace.log
```
And I get the following highest Versions:
```
libavcodec.so.61
libavformat.so.61
```
So it should be possible to switch to `ffmpeg_7-headless`

**Comment by Atemu** [comment]
Please just send a PR then; this one is old enough to walk.

**Pull request merged**
